src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs

```diff
@@ -224,11 +248,28 @@ async Task<CacheResult> ProcessCacheRequest(CacheRequest request)
                     }
                 }
 
-                if (_projectCacheDescriptor.VsWorkaround)
+                // TODO: remove after we change VS to set the cache descriptor via build parameters.
+                // VS workaround needs to wait until the first project is evaluated to extract enough information to initialize the plugin.
+                // No cache request can progress until late initialization is complete.
+                if (_projectCacheDescriptor.VsWorkaround &&
+                    Interlocked.CompareExchange(
+                        ref LateInitializationForVSWorkaroundCompleted,
+                        new TaskCompletionSource<bool>(),
+                        null) is null)
                 {
-                    // TODO: remove after we change VS to set the cache descriptor via build parameters.
                     await LateInitializePluginForVsWorkaround(request);
+                    LateInitializationForVSWorkaroundCompleted.SetResult(true);
                 }
+                else if (_projectCacheDescriptor.VsWorkaround)
+                {
+                    // Can't be null. If the thread got here it means another thread initialized the completion source.
+                    await LateInitializationForVSWorkaroundCompleted!.Task;
+                }
```

Comment:
I think I'd find this easier to read with a nested if: `if (_projectCacheDescriptor.VsWorkaround)`, then inside `if CompareExchange`.